Improve Security of GraphiQL

danpalmer opened the pull request:
This PR makes the following security improvements to GraphiQL in graphene-django.

 - Stores the GraphQL query in the URL fragment, rather than the query, so that sensitive data won't be logged to web server logs.
 - Moves JS to a separate file to enable stricter Content Security Policies.

graingert commented on graphene_django/templates/graphene/graphiql.html:
```diff
@@ -24,41 +24,13 @@
 </head>
 <body>
   <script>
+
```
?

danpalmer replied:
This was me "boyscouting" to make the code more readable, but I decided it was unnecessary when I re-did these changes.

danpalmer commented:
> Might be better to fill the query into the editor, and only execute it when the button is pressed.

> A further security measure could be to use a warning dialogue for queries loaded from the URL when the submit button is pressed?

I think these are orthogonal concerns, but I will look into doing them.

> What about using a fragment instead?

Good call, I will do this.

danpalmer commented:
@graingert can you re-review please. I think this implements everything you suggested, if I understood correctly. Also special cased browser refreshes so that you don't get prompted to allow the query if you're just refreshing the page (which I frequently do to refresh the schema).

graingert commented on graphene_django/templates/graphene/graphiql.html:
```diff
@@ -100,22 +100,29 @@
+    var isReload = window.performance ? performance.navigation.type === 1 : false;
+    if (Object.keys(parameters).length
+      && !isReload
+      && !window.confirm("An untrusted query has been loaded, continue loading query?")) {
```
Might be worth including an explanation of why running an untrusted query might be a bad idea, and what damage it could cause.

Also something like "if you're not sure click cancel"

This is the Facebook devtools warning for eg:

```
Stop!

This is a browser feature intended for developers. If someone told you to copy and paste something here to enable a Facebook feature or "hack" someone's account, it is a scam and will give them access to your Facebook account.

See https://www.facebook.com/selfxss for more information.
```

danpalmer commented:
@graingert I'm not sure it will. I've updated with my implementation, although I'm not sure that it's perfect.

adamchainz commented on graphene_django/templates/graphene/graphiql.html:
```diff
@@ -23,101 +24,6 @@
-      document.body
-    );
-  </script>
+  <script src="{% static "graphene_django/graphiql.js" %}"></script>
```
Doesn't this imply `graphene_django` relies on `staticfiles` now?

danpalmer replied:
Yes. Do you think that needs documenting? I don’t know how optional it is in Django.

danpalmer replied:
cb87f40

graingert commented on graphene_django/templates/graphene/graphiql.html:
```diff
@@ -23,101 +24,6 @@
-      document.body
-    );
-  </script>
+  <script src="{% static "graphene_django/graphiql.js" %}"></script>
```
If you use single quotes for template tag arguments it will look nicer in the GitHub syntax highlighter

danpalmer replied:
This is the style from the Django docs.

danpalmer replied:
Hmmm ok it is pretty bad. 7e8f6db

danpalmer commented:
@syrusakbary could you take a look at this and let me know what you think?

For what it's worth, the main issue - GraphQL queries being in the URL query-string has bitten us and we'll probably have to disable GraphiQL if we can't find a way to improve the security.

danpalmer commented:
@syrusakbary 

> Thanks to the way the GraphQL view is architected, mutations will not be executed if the request is coming through a GET request (meaning: someone accessing an URL from the browser).

Yeah I noticed this, it's a good decision. How about we always eagerly execute queries, and prompt the user for mutations. I think it's worth being clear about the untrusted nature of the mutations.

> The code for the GraphQL view / GraphiQL integration is highly inspired by how graphql-express works. That's why I'm a bit hesitant to go out of the norm for GraphQL/GraphiQL views, since all frameworks we have adopted a similar approach.

Thanks for this, I'm not familiar with other implementations. I agree that we should try to work in a similar way, although I think the change to using a fragment rather than the query string to store the query is an important security feature - so a place where I think it's worth breaking away from their design.

> we'll probably have to disable GraphiQL if we can't find a way to improve the security.

Also, apologies for this, I received feedback that it could be read as much more hostile than I intended. The query string change is necessary for us to start exposing GraphiQL to other developers internally, so I'm happy to make whatever changes are necessary for this to be merged.

danpalmer commented:
@syrusakbary thanks for all the feedback on this.

I've backed out the changes to query execution in GraphiQL for now as I think they need more thought, or just for the change to be made in GraphiQL itself.

Updated the PR description with the features this now implements. Would be great to get this reviewed and merged!

syrusakbary commented:
Great, will take a time to review it practically tomorrow :)

jplock commented:
This PR will make 
https://github.com/graphql-python/graphene-django/pull/308 more difficult to implement. Is there anyway to combine them?

danpalmer commented:
@jplock ah interesting, I don't think this should make it any more difficult to implement #308 – there will be a very minor merge conflict, but there are review points that need addressing that will cause a rewrite of those lines anyway, so I think I'm going to leave the changes for that PR.

danpalmer commented:
Hey, quick ping on this. I believe all feedback has been addressed and that it's ready for merge.

danpalmer commented:
@syrusakbary any update on this?

danpalmer commented:
I believe this is ready to merge and that I've addressed all review points. If there's something outstanding here please let me know – I'm happy to address! Would be nice to get this merged soon though.

danpalmer commented:
@jkimbo are you able to take a look and merge?

danpalmer commented:
@syrusakbary ping on this. It's been hanging around for a while and I don't want it to bit rot. Also we're scaling up the number of devs working with GraphQL, so we would appreciate the security enhancements here (notably not logging the queries in webserver logs).

danpalmer commented:
@syrusakbary @patrick91 @grantmcconnaughey @pizzapanther @spockNinja @RadoRado

Paging contributors. Apologies for the mentions but we've tried pinging and bumping this and haven't get anywhere so far.

It would be a real shame for `graphene-django` to bitrot and stop getting updates. We're investing heavily in it and would love to contribute, but PRs taking months to get merged discourages us from doing so.

This PR is good to go, it's been reviewed, all criticism so far has been addressed (sufficiently I believe). We're more than happy to fix any additional code review points, but otherwise it would be great to get clarity on why this isn't being merged, especially as it's addressing a security vulnerability.

patrick91 commented:
I can approve but I don't have merge permissions, unfortunately

danpalmer commented:
It seems like we have quite high bus factor on this repo. Who has merge permissions? Can they be extended to other frequent contributors like @patrick91 or @jkimbo?

danpalmer commented:
@patrick91 it looks like you do have merge permissions? GitHub requires "1 approving review by reviewers with write access", and it's listing you under that.

pizzapanther commented:
Even if you have GitHub permission what about pypi to publish it?

danpalmer commented:
@pizzapanther merging would at least prevent this PR from bitrotting and would allow people to target the master branch in their requirements if they really wanted to, so I think it would be an improvement, but yes PyPI access would be ideal.

danpalmer commented:
Ah damn, sounds like GitHub PR review requirements and protected branches aren't fully compatible. Thanks for checking though!

danpalmer commented:
@syrusakbary Ping on this. Would be good to talk about the process for PRs getting merged and releases getting shipped in the long run, it would be a real shame for this package to be unmaintained.

syrusakbary commented:
Hi @danpalmer, sorry for the long delay here.
I'm now very immersed in a new venture and takes almost all my time.

I'm working on laying out a plan for Graphene so things like this can be reviewed and merged without my explicit approval, and the project can continue growing at a healthy pace.

pizzapanther commented:
We should call the person who did flatmap-stream on npm. Think he likes to take over projects. 🤣

patrick91 commented:
> Hi @danpalmer, sorry for the long delay here.
> I'm now very immersed in a new venture and takes almost all my time.
> 
> I'm working on laying out a plan for Graphene so things like this can be reviewed and merged without my explicit approval, and the project can continue growing at a healthy pace.

finally!

syrusakbary commented:
Ok, I'm starting to roll the plan.

Please take a look on the main issue in Graphene to provide input and feedback into how you think Graphene can be governed in the long term :)

https://github.com/graphql-python/graphene/issues/884

adamchainz commented:
@jkimbo thanks!
